src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentNodeWriter.cs

```diff
@@ -200,6 +209,18 @@ public override void WriteComponentTypeInferenceMethod(CodeRenderingContext cont
                     p.Add(($"__seq{p.Count}", typeName, $"__arg{p.Count}"));
                 }
 
+                foreach (var capture in node.Component.SetKeys)
+                {
+                    // By declaring the type as "object", this unfortunately means
+                    // we're going to box any value-typed key values before calling
+                    // .GetHashCode() on them. Theoretically we could avoid this by
+                    // adding more generic type params, but that would impact a lot
+                    // the rest of the generic inference implementation. Since the
+                    // combination of generic type inference with keys is reasonably
+                    // uncommon, and one more boxing is not wildly expensive, accept it.
+                    p.Add(($"__seq{p.Count}", "object", $"__arg{p.Count}"));
```

Comment:
It looks like you planned to always treat key as an integer/hash anyway. If that's actually desirable you could do that here. However, I think boxing in this case is fine anyway. 

I agree that adding more generic type parameters would be problematic.

Reply:
I'm backing away from the idea of storing the key as an `int` and using `GetHashCode` for non-int values.

I think now the right thing to do will be to store it as an `object`, so comparisons are based on object identity for reference types (and equality for value types, albeit boxed).

Considerations behind this:

 * **1. Boxing**: Common cases are the key being a model object instance, an `int` primary key value, a `string` primary key value, a `Guid` primary key value. Storing a hash code means no boxing in any case; storing an `object` means boxing in the case of a value-typed primary key (typically `int`). So this is a point in favour of `GetHashCode`.
 * **2. Number of hash computations**: If we store as a hash code, that only has to be computed once per key. If we store as an `object`, then since the matching algorithm uses a dictionary, we have to re-compute `GetHashCode` about twice as often because it's needed for each dictionary read/write. So this is a point in favour of `GetHashCode`.
 * **3. Key collisions**: We have to be tolerant of key collisions because there may be cases where developers can't avoid them (e.g., if they *are* using hash codes as keys, possibly via `key="something.GetHashCode()"` to allow preservation even if `something` got recreated via a new DB query). But the behavior won't be ideal when there are collisions, because we'll have to ignore all but the first usage of a given key value - it's not meaningful to "guarantee preservation of associations" based on non-unique keys. As such, it will be far easier to understand if the matching is based on object identity (for reference types) or equality (for value types), which we get automatically if we store as `object`. If developers *want* hashcode-based keys, they can still always use `key="something.GetHashCode()"`. So this is a point in favour of storing the object instance, not its hash code.

For point (1), I don't think the boxing of int keys is too bad. We're talking about 10-100 items in typical lists, and they don't update faster than humans can make sense of the updates.

For point (2), I think doubling the number of `GetHashCode` calls will be insignificant.

For point (3), I think this is the most important thing, as it actually affects visible behavior and how comprehensible developers will find the system. People will expect matching based on reference-equality. If we don't do that, then it's just a matter of time until random hash code collisions produce near-impossible-to-repro variations in behavior.

Technically we *could* avoid boxing by storing both value-type keys and reference-type keys separately. However I don't want to bloat `RenderTreeFrame` by an extra 8 bytes per frame for every single frame for this, given how I don't think anyone will be able to observe the boxing costs since usage of `key` will only be occasional.

Reply:
I agree totally with the rationale for point 3 btw. Humans are bad at understanding hash codes.

Reply:
I've removed the comment here now since it won't apply under the new "keys are objects" regime.